cpp/src/parquet/page_index.cc

```diff
@@ -0,0 +1,188 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "parquet/page_index.h"
+#include "parquet/encoding.h"
+#include "parquet/exception.h"
+#include "parquet/schema.h"
+#include "parquet/statistics.h"
+#include "parquet/thrift_internal.h"
+
+#include "arrow/util/unreachable.h"
+
+#include <limits>
+#include <map>
+#include <numeric>
+
+namespace parquet {
+
+namespace {
+
+template <typename DType>
+void Decode(std::unique_ptr<typename EncodingTraits<DType>::Decoder>& decoder,
+            const std::string& src, typename DType::c_type* dst) {
+  decoder->SetData(/*num_values=*/1, reinterpret_cast<const uint8_t*>(src.c_str()),
+                   static_cast<int>(src.size()));
+  decoder->Decode(dst, /*max_values=*/1);
+}
+
+template <>
+void Decode<ByteArrayType>(std::unique_ptr<ByteArrayDecoder>&, const std::string& src,
+                           ByteArray* dst) {
+  DCHECK_LE(src.size(), static_cast<size_t>(std::numeric_limits<uint32_t>::max()));
+  dst->len = static_cast<uint32_t>(src.size());
+  dst->ptr = reinterpret_cast<const uint8_t*>(src.data());
+}
+
+template <typename DType>
+class TypedColumnIndexImpl : public TypedColumnIndex<DType> {
+ public:
+  using T = typename DType::c_type;
+
+  TypedColumnIndexImpl(const ColumnDescriptor& descr,
+                       const format::ColumnIndex& column_index)
+      : column_index_(column_index) {
+    size_t num_non_null_pages = std::accumulate(
+        column_index_.null_pages.cbegin(), column_index_.null_pages.cend(), 0U,
+        [](size_t num_non_null_pages, bool null_page) {
+          return num_non_null_pages + (null_page ? 0U : 1U);
+        });
+    if (num_non_null_pages != 0U) {
+      min_values_.reserve(num_non_null_pages);
+      max_values_.reserve(num_non_null_pages);
+      // Decode min and max values into a compact form (i.e. w/o null page)
+      auto plain_decoder = MakeTypedDecoder<DType>(Encoding::PLAIN, &descr);
+      T value;
+      for (size_t i = 0; i < column_index_.null_pages.size(); ++i) {
+        if (!column_index_.null_pages[i]) {
+          non_null_page_indices_.emplace_back(static_cast<int32_t>(i));
```

Comment:
Before downcasting here we should probably check above that the size doesn't go above the limit (unlikely, but who knows).